tactics/dragon.c

```diff
@@ -0,0 +1,656 @@
+ * (board_is_one_point_eye() ones can become false later ...) */
+static bool
+is_real_one_point_eye(struct board *b, coord_t to, enum stone color)
+{
+	return (board_is_eyelike(b, to, color) &&
+		real_eye_endpoint(b, to, color));
+}
+
+static bool
+is_real_two_point_eye(struct board *b, coord_t to, enum stone color, coord_t *pother)
+{
+	if ((neighbor_count_at(b, to, color) +	    
+	     neighbor_count_at(b, to, S_OFFBOARD)) != 3)
+		return false;
+	coord_t other = pass;
+	foreach_neighbor(b, to, {	/* Find the other point ... */
+			if ((board_at(b, c) == S_NONE ||
+			     board_at(b, c) == stone_other(color)) &&
+			    (neighbor_count_at(b, c, color) +	    
+			     neighbor_count_at(b, c, S_OFFBOARD)) == 3) {
+				other = c;
+				break;
+			}
+		});
+	*pother = other;
+	
+	return (!is_pass(other) && 
+		real_eye_endpoint(b, to, color) && 
+		real_eye_endpoint(b, other, color));
+}
+
+struct safe_data {
+	int *visited;
+	int *eyes;
+};
+
+static int
+count_eyes(struct board *b, enum stone color, coord_t lib, void *data)
+{	
+	struct safe_data *d = data;
+	if (d->visited[lib])  /* Don't visit big eyes multiple times */
+		return 0;
+
+	if (is_real_one_point_eye(b, lib, color))  {
+		// fprintf(stderr, "real eye: %s\n", coord2sstr(lib, b));
+		if (++(*d->eyes) >= 2)
+			return -1;
+		return 0;
+	}
+		
+	coord_t other = pass;
+	if (is_real_two_point_eye(b, lib, color, &other))  {
+		// fprintf(stderr, "two-point eye: %s\n", coord2sstr(lib, b));
+		d->visited[other] = 1;
+		if (++(*d->eyes) >= 2)
+			return -1;
+		return 0;
+	}
+		
+	/* TODO check shape ... */
+	int area_size = big_eye_area(b, color, lib, d->visited);
+	if (area_size) {
+		*d->eyes += (area_size > 7 ? 2 : 1);
+		if (*d->eyes >= 2)
+			return -1;
+		return 0;
+	}
+	
+	return 0;
+}
+
+bool
+dragon_is_safe_full(struct board *b, group_t g, enum stone color, int *visited, int *eyes)
+{
+	struct safe_data d = { .visited = visited, .eyes = eyes };
+	foreach_lib_in_connected_groups(b, color, g, count_eyes, &d);
+	return (*eyes >= 2);
+}
+
+bool
+dragon_is_safe(struct board *b, group_t g, enum stone color)
+{
+	int visited[BOARD_MAX_COORDS] = {0, };
+	int eyes = 0;
+	return dragon_is_safe_full(b, g, color, visited, &eyes);
+}
+
+static int
+count_libs(struct board *b, enum stone color, coord_t c, void *data)
+{	
+	int *libs = data;
+	(*libs)++;  return 0;
+}
+
+int
+dragon_liberties(struct board *b, enum stone color, coord_t to)
+{
+	int libs = 0;	
+	foreach_lib_in_connected_groups(b, color, to, count_libs, &libs);
+	return libs;
+}
+
+
+static int
+dragon_at_handler(struct board *b, enum stone color, group_t g, void *data)
+{
+	group_t *d = data;
+	*d = (*d > g ? *d : g);  return 0;		
+}
+
+group_t
+dragon_at(struct board *b, coord_t to)
+{
+	group_t g = group_at(b, to);
+	group_t d = 0;
+	enum stone color = board_at(b, to);
+	if (!g)
+		return 0;
+	
+	foreach_connected_group(b, color, to, dragon_at_handler, &d);
+	return d;
+}
+
+
+#define		GAP_LENGTH        4
```

Comment:
So we are looking for 2-stone gap but GAP_LENGTH is 4.  It'd be nice to have a single-sentence comment explaining this...